### ISSUE TITLE
Add a new cygwin target

### ISSUE THREAD

**Issue by mintty** [rust-lang/rust#79854]
According to the sparse build instructions available, rust should be built by running x.py.
After a bunch of secondary downloads, removing obstacles with curl, it fails at:
error: manifest path `/cygdrive/c/tmp/rust/src/bootstrap/Cargo.toml` does not ex
ist
failed to run: /cygdrive/c/tmp/rust/build/i686-pc-windows-gnu/stage0/bin/cargo build --manifest-path /cygdrive/c/tmp/rust/src/bootstrap/Cargo.toml

I found https://github.com/rust-lang/rust/issues/5526#issuecomment-17460380 but actually any package that builds on Linux should generally also be buildable in the cygwin Posix environment.

**Comment by jyn514** [comment]
@mintty something strange is going on in your environment: `bootstrap/Cargo.toml` is part of the repository: https://github.com/rust-lang/rust/blob/master/src/bootstrap/Cargo.toml. Did you modify or rename it?

**Comment by mintty** [comment]
No, the file is there.
-rw-rw-r--+ 1 towo Kein 1020 Dec  8 21:23 /cygdrive/c/tmp/rust/src/bootstrap/Cargo.toml

**Comment by qiqjiao** [comment]
What's the status now? Seems the cygwin support was asked many years ago in other threads. I'm installing a python package in cygwin, that thing depends on cryptography, which depends on a rust compier.

**Comment by mintty** [comment]
First attempt (python x.py):
...
Building rustbuild
error: manifest path `/cygdrive/c/tmp/rust/src/bootstrap/Cargo.toml` does not exist
failed to run: /cygdrive/c/tmp/rust/build/i686-pc-windows-gnu/stage0/bin/cargo build --manifest-path /cygdrive/c/tmp/rust/src/bootstrap/Cargo.toml

but the file does exist

Second attempt (./x.py):
...
   Compiling cc v1.0.73
error: linker `i686-w64-mingw32-gcc` not found

Installed package  mingw64-i686-gcc-core

Third attempt (./x.py):
...
error: linking with `i686-w64-mingw32-gcc` failed: exit code: 1
...
error: could not compile `syn` due to previous error
error: could not compile `serde_derive` due to previous error
error: could not compile `proc-macro2` due to previous error
error: could not compile `winapi` due to previous error
error: could not compile `typenum` due to previous error
failed to run: C:\tmp\rust\build\i686-pc-windows-gnu\stage0\bin\cargo.exe build--manifest-path C:\tmp\rust\src/bootstrap/Cargo.toml

**Comment by mati865** [comment]
It builds fine in MSYS2 so there should be no problem with Cygwin.

> error: manifest path `/cygdrive/c/tmp/rust/src/bootstrap/Cargo.toml` does not exist

This is unexpected, seems Cygwin path was passed instead of Windows path.
Maybe because of Python? In that case using native Windows Python would solve the problem.


> error: linking with `i686-w64-mingw32-gcc` failed: exit code: 1
> ...
> error: could not compile `syn` due to previous error

This output doesn't contain the root cause of the failure. You should look higher up in the terminal (or even paste whole output somewhere) and if there is nothing interesting re-run with `-vv` for more verbose output.

**Comment by mintty** [comment]
> Maybe because of Python? In that case using native Windows Python would solve the problem.

Maybe, but for building a cygwin package a cygwin toolchain is essential.

**Comment by mati865** [comment]
Ah, I know what it is. I think all Cygwin packages do use UNIX style paths but native Windows applications use only Windows style paths. So there are two problems here:
- Cygwin tools pass UNIX style paths to Rust but it cannot understand them
- Rust then passes Windows paths to Cygwin tools (like the linker) which they cannot understand AFAIK

So to do it that way you'd need to convert paths using `cygpath` on the boundaries between Cygwin and Win32 apps.

Side question: do you want Cygwin host+target or Cygwin host but targeting only MinGW?
For the first option you cannot use MinGW targets. You would need to create Cygwin target triple called probably `x86_64-pc-windows-cygnus` or something like that.

**Comment by mintty** [comment]
I'm only interested in cygwin target. I have no capacity to work on your build, anyway. I was just reporting as I think (quoting from above)
> any package that builds on Linux should generally also be buildable in the cygwin Posix environment.

**Comment by mati865** [comment]
> any package that builds on Linux should generally also be buildable in the cygwin Posix environment.

This statement is not true when it comes to the compilers though.

**Comment by qiqjiao** [comment]
> This statement is not true when it comes to the compilers though.
Compilers translate source code to machine code. It doesn't involve low level OS specific things like the device drivers. I don't see why this particular compiler is special. No official support of cygwin rust compiler package is understandable, but not even be able to build from source code is hard to understand.

**Comment by mati865** [comment]
> Compilers translate source code to machine code. It doesn't involve low level OS specific things like the device drivers.

No matter if its C++ or Rust, std library is all about using OS specific code to give users cross-platform and easy to use abstractions though.

**Comment by mati865** [comment]
> > Compilers translate source code to machine code. It doesn't involve low level OS specific things like the device drivers.
> 
> No matter if its C++ or Rust, std library is all about using OS specific code to give users cross-platform and easy to use abstractions though.

> not even be able to build from source code is hard to understand.

All the errors in this issue are about the paths, you can take any Win32 native application and it won't be able to take or output UNIX paths used by Cygwin because Win32 API itself doesn't support these paths.

**Comment by mintty** [comment]
> All the errors in this issue are about the paths,

I guess something in the Rust build scripts is confused by the cygwin environment, misinterpreting it as a Windows environment and thus using wrong path syntax. It should be easy to find that place for those familiar with the build scripts.

**Comment by mintty** [comment]
This composes os.path.join(self.rust_root, "src/bootstrap/Cargo.toml") which works fine in cygwin, with cygwin python.
So probably elf.rust_root contains a Windows path already.

**Comment by mati865** [comment]
No, it's created using relative paths to Python file and then normalised: https://github.com/rust-lang/rust/blob/9fb32dc924653e35950f17c8d91793c9ca983d03/src/bootstrap/bootstrap.py#L846

**Comment by DWesl** [comment]
Cygwin often (but not always) can understand and use Windows-style paths just fine; the problems I've run into have mostly been of the "Windows executable doesn't understand Cygwin-style path" variety (although that's started to change recently; I'm seeing more `/path/to/pwd/C:/path/to/somewhere`-related failures).  

The first question that occurs to my is why the build wants to use `i686-w64-mingw32-gcc` as the linker, since that will generate 32-bit native Windows applications, not Cygwin applications.  Searching for Cygwin in the codebase, I found
https://github.com/rust-lang/rust/blob/41ad4d9b2dbb895666337d162eda52619a6056db/src/bootstrap/bootstrap.py#L268-L289
and
https://github.com/rust-lang/rust/blob/fee3a459dd6aba8e34a5b99f0fbcb4218a1e2401/src/ci/shared.sh#L47-L49
which seems to indicate that the rust source pretends that, say `i686-pc-cygwin` on a 64-bit Windows is actually `x86_64-w64-mingw32`, that is, the code base as-is explicitly disallows attempting to create a true Cygwin-native Rust; it will create either a Windows-native version of Rust or maybe an MSYS version of Rust, both already available from https://forge.rust-lang.org/infra/other-installation-methods.html#other-ways-to-install-rustup.

To build an actual Cygwin version of Rust, one would first need to change the code in both the above places so the code sees Cygwin as a thing distinct from MSVC, MinGW, and MSYS, then figure out which of the tests for "is this Windows?" are actually for:
- are executables PE/COFF or ELF? (does specifying `rpath` forbidden or required?)
- does the linker need to be able to resolve all symbols, or can that wait until run time?
- does the search for shared libraries at executable startup generally happen via `PATH` or `LD_LIBRARY_PATH`?
- do functions used in shared libraries need `__declspec(dllexport/dllimport)` to work properly or no?
- is the end-of-line marker CRLF or just LF?
- will Windows-native executables run if passed only relative paths, or only in an emulator/VM?
- are Windows API functions available (if perhaps persnickety) or no?
- are the type sizes LLP64 or LP64?  (is `long` 32 bits or 64 on a 64-bit platform?)
- are POSIX functions basically unavailable, or only gated behind visibility macros (`-D_POSIX_C_SOURCE` and friends`)?
- is the default file encoding UTF-16 or UTF-8?
- are colons (`:`) banned file names (except for right after the drive letter) or mostly fine?
- is the path separator in `PATH` a semi-colon (`;`) or a colon (':`)?

Some of these questions have the same answers in MSVC and Cygwin, so that treating them the same way makes sense; others have different answers in MSVC and Cygwin, and must be treated differently.  I've run into both kinds of problems in different packages.

It's probably possible to disentangle those questions; as a starting point, the API (C-level and above) is closest to BSD (not usually too far from Linux aside from not using `glibc`; it looks like Rust already fixed those compatibility problems), while the linking semantics (link-time and run-time) are closer to Windows.

In short, earlier comments suggest OP wants to build a Rust compiler that is itself a Cygwin executable that generates Cygwin executables; the alternative is that they want to use Cygwin to provide a build platform to (cross-)compile their own version of the Rust compilers available from `rust-lang.org` (a Windows-native executable that generates Windows-native executables).  The current build system seems to bake in the second one of those as the only supported option, but runs into problems with path conversion (because it thinks it's not cross-compiling).

**Comment by mati865** [comment]
> The first question that occurs to my is why the build wants to use i686-w64-mingw32-gcc as the linker, since that will generate 32-bit native Windows applications, not Cygwin applications.

As I already said there is no Cygwin target right now so you have to add it first to build Rust as Cygwin application. 

As for the build system detection or MinGW you can override it with `--build <triple>`, `--host <triple>` and `--target <triple>`. After Cygwin targets are added.

**Comment by jschultz** [comment]
I'd like to have a go building Rust for Cygwin. I see the bootstrapping process begins with x.py, but even it seems to start by downloading some pre-built binaries. Can anyone show me full bootstrap instructions from nothing but source code?

**Comment by jyn514** [comment]
@jschultz https://rustc-dev-guide.rust-lang.org/building/new-target.html#adding-a-new-target

**Comment by jschultz** [comment]
Sorry but any chance of a little more hand-holding? Whatever I do x.py seems to start by downloading `rust-std-beta-i686-pc-windows-gnu.tar.xz`. What I think I need to do is to build something like `rust-std-beta-i686-cygwin.tar.xz`. From my experience building under cygwin doing so involves little more than doing a Linux build under cygwin, then dealing with various random incompatibilities.

**Comment by jyn514** [comment]
@jschultz this target doesn't exist yet, so you need to cross compile to it from an existing target.

**Comment by DWesl** [comment]
> DWEST asked these questions above., thought I would take a stab at them:

As you may have noticed, the Windows option is the first given in each question except the last: the reason I posed the question that way was that rust, at the time, assumed that when I executed `./configure && make` on Cygwin, I wanted a `x86_64-w64-mingw32` build.  This is an okay starting point if the main library is built on the assumption that everything else is Linux and the main question to be answered is what version of gnu libc is available.  

I have seen configuration tie the answers to all of those questions to a simple binary "Does the platform name contain the string 'win'?" or "Is the platform Windows?".  Cygwin complicates the issue, with some questions answered the same as on Windows (the first three questions), and others closer to their answers on BSD (most of the rest, except the declspec and Windows API ones).
> 
> > are Windows API functions available (if perhaps persnickety) or no?
> 
> Not unless the Windows API funcs are defined in POSIX or some compatibility lib.

Apparently Cygwin defines these in a compatibility header, which you can technically use if you are very, very careful about types because
> 
> > are the type sizes LLP64 or LP64? (is long 32 bits or 64 on a 64-bit platform?)
> 
> Is that something that POSIX specifies? Usually config scripts test things like that.

Windows is LLP64, while Linux is LP64 so `sizeof(long)` won't match assumptions.
I suspect this is outside the scope of POSIX. 
> 
> > is the default file encoding UTF-16 or UTF-8?
> 
> That's not specified by POSIX, but I would lean toward 8-bits/char. No encoding at the POSIX layer (needs to be done by applications needing something more than ASCII). It would be like asking what is linux's default file encoding. Answer -- 8-bit characters.

Windows defaults to UTF-16, Cygwin seems to default to UTF-8 these days, though this used to vary by region.
> 
> Generally I can take 'rpm's from linux and build them in Cygwin with no changes (presuming they put binaries in places like '/usr/bin'. 

Usually, yes, but not always.  Many packages assume that linking some library will pull in it's dependent library's symbols at runtime, so they don't need to bother with linking those.  On Windows and Cygwin, this is an error; the dependent library must be specified at link time.  A couple will try some ELF-specific thing and crash, or try to use functions from GNU Libc and also crash.

Many packages, when first porting from Linux to Windows, will assume for simplicity that answering one of these questions will provide the answer for all the others.  Porting to Cygwin involves the process of decoupling those answers into two to three groups.  (Requiring all symbols be resolvable at runtime and refusing to recognize `rpath` as a concept doesn't imply a thing about whether tests should expect CRLF line endings or 32-bit long on 64-bit platforms).

**Comment by Astara** [comment]
On 2023/03/27 17:15, DWesl wrote:
>
>         are the type sizes LLP64 or LP64? (is long 32 bits or 64 on a
>         64-bit platform?)
>
>     Is that something that POSIX specifies? Usually config scripts
>     test things like that.
>
> Windows is LLP64, while Linux is LP64 so |sizeof(long)| won't match 
> assumptions.
> I suspect this is outside the scope of POSIX.
>
---
    But those sizes have never been constants -- every source needs to 
test for
sizes if they rely on them.
>
> Windows defaults to UTF-16, Cygwin seems to default to UTF-8 these 
> days, though this used to vary by region.
>
---
    Internally Windows defaults to UCS-2LE, not UTF-16.  UCS-2LE 
allocates 2 bytes/character
but no way to extend beyond 16-bits -- that came later in user compat 
libs outside the
kernel.  Windows is mostly UTF-8 to be compatible with the web.  It's 
rare to find
a website hosted on windows in UTF-16.  What can be worse is that many 
people have
the mistaken assumption that USC-2 will always include a BOM so LE can 
be differentiated
from BE.  Files coming directly out of the kernel (log files, 
especially) often
don't have BOM's.  It's smarter to assume LE because of Windows being 
native to
LE processors, though some BE processors have been supported over the years.

    But none of that has anything to do with POSIX compatibility.

>     Generally I can take 'rpm's from linux and build them in Cygwin
>     with no changes (presuming they put binaries in places like
>     '/usr/bin'.
>
> Usually, yes, but not always.
>
---
Nothing is ever "always", otherwise it is identical.
>
> Many packages assume that linking some library will pull in it's 
> dependent library's symbols at runtime, so they don't need to bother 
> with linking those. On Windows and Cygwin, this is an error; the 
> dependent library must be specified at link time. A couple will try 
> some ELF-specific thing and crash, or try to use functions from GNU 
> Libc and also crash.
>
---
    Of course, windows is COFF, not ELF.  But you can use DLL's (Delayed 
Load Libraries), to
have symbols linked at runtime (even if they had to be resolved at link 
time).

> Many packages, when first porting from Linux to Windows, will assume 
> for simplicity that answering one of these questions will provide the 
> answer for all the others.
>
---
    That would be a naive assumption. 
>
> Porting to Cygwin involves the process of decoupling those answers 
> into two to three groups. (Requiring all symbols be resolvable at 
> runtime and refusing to recognize |rpath| as a concept doesn't imply a 
> thing about whether tests should expect CRLF line endings or 32-bit 
> long on 64-bit platforms).
>
---
    That's why Gnu's 'configure' was invented -- it tested each item so 
programs using it
could be run on a large number of target platforms.  Now people try to 
make due with
less precise auto-config tools that make invalid assumptions by trying 
to group all target
platforms into 2-3 groups -- which just won't work in reality.

**Comment by ookiineko** [comment]
> if someone want to start a Rust port for Cygwin, my previous attempt may be somewhat useful

it has some initial libc and STD support (untested probably it has tons of typos, since ctest won't work properly without stack unwinding), some crates were patched, and i have got helloworld and some simple program like ripgrep to work...

but that was a little bit outdated now since i wasn't working on it for a while

i may back to upstreaming it after i can fix LLVM build...

**Comment by ookiineko** [comment]
> FYI it's possible to add a cygwin _target_ without getting rustc _itself_ to build on cygwin. That wouldn't let you run rustc within cygwin, but it would let you run `cargo build --target x86_64-pc-windows-cygwin` from a mingw or MSVC toolchain.

i understand that, i was trying to cross compile the rustc from a Fedora Linux host

**Comment by ookiineko** [comment]
> but it would let you run `cargo build --target x86_64-pc-windows-cygwin` from a mingw or MSVC toolchain.

i couldn't find any Cygwin cross toolchain on MinGW or MSVC... maybe I should build it myself

i have tried MSYS2 and directly run the Cygwin toolchain from Cygwin sysroot in MinGW environment (since they're all win32 runnable)

but the path handling caused the build to fail

**Comment by ookiineko** [comment]
> Member

currently I am able to do `cargo build --target x86_64-pc-cygwin` from a Fedora host

**Comment by workingjubilee** [comment]
Some people seem to misunderstand this issue as something that is going to be addressed  simply because there has been an issue opened for it. This issue by itself will not provoke any action from the Rust teams on adding this target, as we would require a maintainer with specialized knowledge regarding this target to add it in the first place. So please follow the [target tier policy](https://doc.rust-lang.org/rustc/target-tier-policy.html) for adding a new target. It does not require an issue on this issue tracker to do so.

As the ancient proverb goes,
"PRs welcome!"

Closing.

**Comment by JavaScriptDude** [comment]
I sure hope a solution can be found for this as Cygwin will be in a slow degrade as a result of the underlying issue. `Cryptography` isn't the only CPython based  library that uses rust and I keep seeing more libraries pop up as time progresses; great for Rust but bad for Cygwin.

Anybody make a write-up on this; seems like a good hacker news post ;^)

**Comment by ookiineko** [comment]
currently the main issue is Cygwin lacks of LLVM maintainers and the related packages are outdated, this should get fixed with some efforts. a rough example is [here](https://github.com/ookiineko-cygpkg/llvm) (9.0 -> 12.0, Libc++ is currently buggy)

cygwin-apps seems to require Ocaml binding to work when upgrading the packages, which involves reviving not just the LLVM packages, but also the Ocaml ones. that's quite a lot of efforts.

after that some initial work for adding the Cygwin target for Rust/STD support/libc crate is done in [here](https://github.com/orgs/ookiineko-cygport/repositories) (Rust 1.67.0 nightly, because that's the latest version when i started to work on it, currently waiting for Cygwin LLVM updates before i decide to port newer Rust).

also see the related thread i started at the cygwin-apps mailing list (July, Aug, Sep): https://cygwin.com/pipermail/cygwin-apps/2023-July/043048.html

i do want to share my journey porting Rust and cross-compiling rust programs for running on Cygwin, but it's buggy and full of hack right now, and i don't have a blog so i don't know when i will do that...

**Comment by ookiineko** [comment]
> a rough example is [here](https://github.com/ookiineko-cygpkg/llvm) (9.0 -> 12.0, Libc++ is currently buggy)

finally now i have managed to get llvm15 toolchain patched (including clang and libc++) for cygwin and get it to do sth

so recently i will try to see if i can get a host rustc cross-compiled for cygwin

> currently the main issue is Cygwin lacks of maintainers for LLVM packages

also there is someone in the mailing list planning to become new maintainer of the llvm pkgs
https://cygwin.com/pipermail/cygwin-apps/2024-January/043374.html

**Comment by Kreijstal** [comment]
> > a rough example is [here](https://github.com/ookiineko-cygpkg/llvm) (9.0 -> 12.0, Libc++ is currently buggy)
> 
> finally now i have managed to get an experimental llvm15 toolchain patched (including clang and libc++) for cygwin and get it to do sth
> 
> btw i have now cross-compiled host rustc for cygwin, it starts, but currently not good enough to compile programs (as all kinds of small bugs are starting to show up
> 
> i guess for now i will be focusing on trying to pass the libc-test first
> 
> > currently the main issue is Cygwin lacks of maintainers for LLVM packages
> 
> also there seems to be a guy in the mailing list who is planning to become new maintainer of the llvm pkgs https://cygwin.com/pipermail/cygwin-apps/2024-January/043374.html

did you manage create a rust toolchain for cygwin?

**Comment by ookiineko** [comment]
> did you manage create a rust toolchain for cygwin?

for now i have a working cross toolchain on linux, that is able to cross-compile some cygwin programs.

for a rust toolchain on cygwin, i haven't yet got it working. although i have `rustc.exe`, `cargo.exe` and several stuffs successfully compiled and starting, it is not fully functioning yet

**Comment by Kreijstal** [comment]
> > did you manage create a rust toolchain for cygwin?
> 
> for now i have a working cross toolchain on linux, that is able to cross-compile some cygwin programs.
> 
> for a rust toolchain on cygwin, i haven't yet got it working. although i have `rustc.exe`, `cargo.exe` and several stuffs successfully compiled and starting, it is not fully functioning yet
> 
> i have a weird issue with LLVM initialization in rustc, it jumps to `0x0` from `std::call_once` (libstdc++ function) when doing `llvm::initializeCore` where it should jump to a function inside LLVM, the address of the function is stored in a TLS variable, and Cygwin uses emutls, that's all I know so far. This doesn't happen outside rustc, it also doesn't happen to other programs using LLVM (like Clang) on Cygwin.
> 
> I'm still trying to figure out why. I don't have a lot of debugging experience with gdb, but i will try to learn more and diagnose

can you document your steps, maybe someone can help

**Comment by ookiineko** [comment]
> can you document your steps, maybe someone can help

sure, im currently setting up a new set of cygwin cross compilers at [arch-cygwin](https://github.com/ookiineko/arch-cygwin) (since the fedora cygwin i was using is inactive), i will write down my journey and a build guide in details (probably make it a blog) once im ready

**Comment by Kreijstal** [comment]
> > can you document your steps, maybe someone can help
> 
> sure, im currently setting up a new set of cygwin cross compilers at [arch-cygwin](https://github.com/ookiineko/arch-cygwin) (since the fedora cygwin i was using is inactive), i will write down my journey and a build guide in details (probably make it a blog) once im ready

that's great  to hear

**Comment by workingjubilee** [comment]
I should note we'd probably love to have more maintainer support for the `*-pc-windows-gnullvm` targets that use the UCRT instead. That way we could promote them from tier 3.

**Comment by mati865** [comment]
> I should note we'd probably love to have more maintainer support for the `*-pc-windows-gnullvm` targets that use the UCRT instead. That way we could promote them from tier 3.

They were accepted as tier 2: https://github.com/rust-lang/rust/pull/121712 but they are blocked on `compiler_builtins` being currently broken for `i686-pc-windows-gnu` (the one without `llvm` suffix): https://github.com/rust-lang/rust/pull/121712

**Comment by ookiineko** [comment]
> it would let you run `cargo build --target x86_64-pc-windows-cygwin` from a mingw or MSVC toolchain.

i have made one for mingw host in the last few days: https://github.com/ookiineko-cygport/rust/releases/test-build

![image](https://github.com/rust-lang/rust/assets/103345643/02927e5a-70f9-494b-9f08-11922ea2b96f)

so now its easier to test and debug this thing (cross compiling from linux host is no longer required

**Comment by riedel** [comment]
@ghost : the gist gives me a 404 , can you reshare it again?

**Comment by Kreijstal** [comment]
> [@ghost](https://github.com/ghost) : the gist gives me a 404 , can you reshare it again?

ghost means the user deactivated the account

**Comment by jeremyd2019** [comment]
It looks like the repositories moved to @ookiineko-cygport at least.

**Comment by Kreijstal** [comment]
yep, we enjoy this https://github.com/ookiineko-cygport/rust

**Comment by Kreijstal** [comment]
now the only thing left is try to bootrstrap it from mrustc

**Comment by mati865** [comment]
It'd be a lot easier to cross compile from another existing target than to bootstrap from mrustc once the necessary changes are merged.
Mrustc doesn't even properly support mingw-w64: https://github.com/thepowersgang/mrustc/blob/95023de01fecee5ba94c4dc50e48ed1579998699/tools/minicargo/os.cpp#L18.

**Comment by Kreijstal** [comment]
> It'd be a lot easier to cross compile from another existing target than to bootstrap from mrustc once the necessary changes are merged. Mrustc doesn't even properly support mingw-w64: https://github.com/thepowersgang/mrustc/blob/95023de01fecee5ba94c4dc50e48ed1579998699/tools/minicargo/os.cpp#L18.

We write pull requests for mrustc or patch it

**Comment by faho** [comment]
Since I've seen people be confused by opening this issue: a cygwin target has been merged with #134999

**Comment by workingjubilee** [comment]
yes. was closed as wontfix. closing as completed.

**Comment by workingjubilee** [comment]
Opened a new issue for "the target works enough to be built as a host", since there's been so many requests for it: https://github.com/rust-lang/rust/issues/137819
